src/windows_native/dev_node.rs

```diff
@@ -0,0 +1,71 @@
+use std::ptr::null_mut;
+use windows_sys::Win32::Devices::DeviceAndDriverInstallation::{CM_Get_DevNode_PropertyW, CM_Get_Parent, CM_LOCATE_DEVNODE_NORMAL, CM_Locate_DevNodeW, CR_BUFFER_SMALL, CR_SUCCESS};
+use windows_sys::Win32::Devices::Properties::DEVPROPKEY;
+use crate::ensure;
+use crate::windows_native::error::{check_config, WinError, WinResult};
+use crate::windows_native::string::U16Str;
+use crate::windows_native::types::DeviceProperty;
+
+#[repr(transparent)]
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct DevNode(u32);
+
+impl DevNode {
+
+    pub fn from_device_id(device_id: &U16Str) -> WinResult<Self> {
+        let mut node = 0;
+        let cr = unsafe {
+            CM_Locate_DevNodeW(&mut node, device_id.as_ptr(), CM_LOCATE_DEVNODE_NORMAL)
+        };
+        check_config(cr, CR_SUCCESS)?;
+        Ok(Self(node))
+    }
+
+    pub fn parent(self) -> WinResult<Self> {
+        let mut parent = 0;
+        let cr = unsafe { CM_Get_Parent(&mut parent, self.0, 0) };
+        check_config(cr, CR_SUCCESS)?;
+        Ok(Self(parent))
+    }
+
+    fn get_property_size<T: DeviceProperty>(self, property_key: *const DEVPROPKEY) -> WinResult<usize> {
```

Comment:
I would argue that the fn needs to be unsafe. A caller could pass in whatever they want as property_key (invalid pointers), and the function would then cause undefined behavior.

One fix would be to make it unsafe.
Another to pass property_key as `&DEVPROPKEY` meaning it must be valid

Reply:
I don't think that there is any safety problem here because the underlying api already seems to verify the argument.
 
Some tests that I ran with an attached debugger:
```
[src\windows_native\device_info.rs:96] dev_node.get_property(&420 as *const i32 as *const DEVPROPKEY) = Err(NoSuchValue)
```
```
[src\windows_native\device_info.rs:96] dev_node.get_property(0x1234567890 as *const DEVPROPKEY) = Err(Config(CR_FAILURE))
```
```
[src\windows_native\device_info.rs:96] dev_node.get_property(null()) = Err(Config(CR_FAILURE))
```